source/agora/node/Ledger.d

```diff
@@ -1067,6 +1071,22 @@ public class Ledger
     {
         return this.enroll_man.addPreimages(preimages);
     }
+
+    /***************************************************************************
+
+        Generate the random seed reduced from the preimages for the current
+        height
+
+        Returns:
+            the random seed if there are one or more valid preimages,
+            otherwise Hash.init.
+
+    ***************************************************************************/
+
+    public Hash getRandomSeed() nothrow @safe
+    {
+        return this.slash_man.getRandomSeed(this.last_block.header.height);
+    }
```

Comment:
This should be in the slashing manager directly and not in the Ledger. The problem is, we shouldn't need to do dependency injection here, because the random seed is derived from pre-images and shouldn't need to be changed. Or am I missing something ?

Reply:
I understand what you mean. But Getting the random seed is needed in unit tests like in `TestLedger` so I've got to make this function overridden which is used as follows.

```
static class ThrowingLedger : Ledger
    {
        // ...

        ///
        public override Hash getRandomSeed() nothrow @safe
        {
            return getTestRandomSeed();
        }
    }
```